tools/qa/simulate_trapdoor_pr.py

```diff
@@ -0,0 +1,237 @@
+        """Dectorator to add section output to function.
+
+        Parameters
+        ----------
+        name : str
+               The function name.
+        """
+        def decorator(fn):
+            """A function that returns the wrapped version of the original function."""
+            def wrapper(*args, **kwargs):
+                """The wrapper with additional printing."""
+                self('BEGIN %s.' % name, indent=2)
+                result = fn(*args, **kwargs)
+                self('END %s.' % name, indent=2)
+                return result
+            return wrapper
+        return decorator
+
+
+log = Log('SIMULATE_TRAPDOOR_PR')
+
+
+def main():
+    """Main program."""
+    # A few general things.
+    args = parse_args()
+    repo = git.Repo('.')
+
+    # Get the QAWORKDIR. Create if it does not exist yet.
+    qaworkdir = os.getenv('QAWORKDIR', 'qaworkdir')
+    if not os.path.isdir(qaworkdir):
+        os.makedirs(qaworkdir)
+
+    # Pre-flight checks.
+    orig_head_name, merge_head_name = run_pre_flight_checks(repo)
+
+    try:
+        make_temporary_merge(repo, merge_head_name)
+        trapdoor_workflow(repo, args.script, qaworkdir, args.skip_ancestor)
+    finally:
+        roll_back(repo, orig_head_name, merge_head_name)
+
+
+def parse_args():
+    """Parse command-line arguments."""
+    parser = argparse.ArgumentParser(description='Simulate trapdoor test locally.')
+    parser.add_argument('script', type=str, metavar='trapdoor', help='Path to trapdoor script.')
+    parser.add_argument('-s', '--skip-ancestor', default=False, action='store_true',
+                        help='Do not run the trapdoor on master and re-use result for '
+                             'ancestor from previous run.')
+    return parser.parse_args()
+
+
+@log.section('pre-flight checks')
+def run_pre_flight_checks(repo):
+    """Run some initial checks before doing anything.
+
+    Parameters
+    ----------
+    repo : git.Repo
+           A repository object from GitPython.
+    """
+    log('Check whether all changes are committed.')
+    # (Can be made better in future version by temporarily auto-committing.)
+    if repo.is_dirty():
+        raise RepoError('Not all changes are committed.')
+
+    log('Check whether master is up to date with origin/master.')
+    remote_refs = git_ls_remote('origin')
+    if remote_refs['refs/heads/master'] != repo.heads.master.object.hexsha:
+        raise RepoError('Master is not up to date.')
+
+    log('Check whether master is currently _not_ checked out.')
+    if repo.active_branch == repo.heads.master:
+        raise RepoError('The master branch is checked out.')
+
+    log('Check whether temporary branch name does not exist yet.')
+    orig_head_name = repo.active_branch.name
+    merge_head_name = '%s-trapdoor-tmp-merge' % orig_head_name
+    if merge_head_name in repo.heads:
+        raise RepoError('The branch %s exists, probably due to earlier failures of this program.')
```

Comment:
Yes, that should be added eventually. For now, i.e. 2.0.1, I've only implemented a minimal conservative version because that was the only thing I needed for fixing a bug. We should add more features to this script to make it more suitable for routine use:
- cleaning up old branch, as you said
- updating the master if needed
- including uncommitted changes automatically.in the temporary merge.
These are features that need to be implemented with care, because they manipulate the local repo even more. I don't want to rush with that before the 2.0.1 release at the risk of introducing bugs that cause irreversible damage.